main/lsp/requests/completion.cc

```diff
@@ -964,24 +964,49 @@ unique_ptr<ResponseMessage> CompletionTask::runRequest(LSPTypecheckerDelegate &t
         auto prefix = (callerSideName == core::Names::methodNameMissing() || !sendResp->funLoc.contains(queryLoc))
                           ? ""
                           : callerSideName.shortName(gs);
-        config.logger->debug("Looking for method similar to '{}'", prefix);
-
-        // isPrivateOk means that there is no syntactic receiver. This check prevents completing `x.de` to `x.def`
-        // (If there is a method whose name overlaps with a keyword, it will still show up as a _method_ item.)
-        auto suggestKeywords = sendResp->isPrivateOk;
-        auto params = SearchParams{
-            queryLoc, prefix,
-            MethodSearchParams{
-                sendResp->dispatchResult,
-                sendResp->totalArgs,
-                sendResp->isPrivateOk,
-            },
-            suggestKeywords,
-            // No receiver means that local variables are allowed here.
-            sendResp->isPrivateOk ? sendResp->enclosingMethod : core::MethodRef{},
-            core::lsp::ConstantResponse::Scopes{}, // constants don't make sense here
-        };
-        items = this->getCompletionItems(typechecker, params);
+        if (prefix == "" && queryLoc.adjust(gs, -2, 0).source(gs) == "::") {
+            // Probably a case like this:
+            //   A::|
+            //   puts
+            // which parses as a method call (A.puts()) instead of a constant.
+
+            auto scopes = core::lsp::ConstantResponse::Scopes{};
+            // We're ignoring the other dispatch components here, because the fast that we saw `::`
+            // in the source means that it's likely a constant lit, not something with a fancy type.
+            // Also since it's a constant lit, it probably has type `T.class_of(...)`, so we want
+            // the represented type for the sake of constant lookup.
+            auto recv = core::Types::getRepresentedClass(gs, sendResp->dispatchResult->main.receiver);
+            if (recv.exists()) {
+                scopes.emplace_back(recv);
+            }
+
+            auto params = SearchParams{
+                queryLoc,
+                prefix,
+                nullopt,           // do not suggest methods
+                false,             // do not suggest keywords
+                core::MethodRef{}, // do not suggest locals
+                scopes,
+            };
+            items = this->getCompletionItems(typechecker, params);
+        } else {
+            // isPrivateOk means that there is no syntactic receiver. This check prevents completing `x.de` to `x.def`
+            // (If there is a method whose name overlaps with a keyword, it will still show up as a _method_ item.)
+            auto suggestKeywords = sendResp->isPrivateOk;
```

Comment:
I think the rationale in the comment or the logic for `suggestKeywords` needs to be adjusted:

https://github.com/sorbet/sorbet/blob/10481e1af8df9913da6e57e5960fee5c736cbf2e/ast/desugar/Desugar.cc#L645-L648

since we don't want `self.de` to complete with keywords.

Also `super` can be `isPrivateOk`, so you may want to check for that specifically too?

Reply:
Thanks. The first one is an existing bug. I'll open a ticket for it, but I don't want to block this PR on fixing it.

https://github.com/sorbet/sorbet/issues/5157

The `super` question is actually subtle, because it gets translated to `<call-with-block>(<self>, :<super>, <blk>)` which then suggests no completion results. So it does it in a roundabout way but it works.